src/transformers/training_args.py

```diff
@@ -477,7 +477,7 @@ class TrainingArguments:
         metric_for_best_model (`str`, *optional*):
             Use in conjunction with `load_best_model_at_end` to specify the metric to use to compare two different
             models. Must be the name of a metric returned by the evaluation with or without the prefix `"eval_"`. Will
-            default to `"loss"` if unspecified and `load_best_model_at_end=True` (to use the evaluation loss).
+            default to `"loss"` if unspecified, `load_best_model_at_end=True`, and `save_strategy!="best"`.
```

Comment:
my 2cs (Disclaimer! I'm not very familiar with the whole scope of the initial change, or reason behind it!): it's a bit hard to read and understand what is going on here and why. E.g. why can't it default to `loss` when save_strategy == best? What is the major difference with the `load_best_model_at_end` (and save_strategy!="best")?

Again, apologies if I'm missing some obvious context here. Please feel free to disregard my comment / question then.

Reply:
I didn't find the place where we set `metric_for_best_model = "loss"` when `save_strategy!=best`. Can you explain a bit why you changed the description ?

Reply:
@shcheklein That was a design decision made here (https://github.com/huggingface/transformers/pull/31817#discussion_r1806713897). It was deemed easier to debug if we don't add a hard-coded value and rather raise an error.

---

@SunMarc Hmm I'm starting to think that maybe the problem is that we're not able to set `load_best_model_at_end = True` when `save_strategy = "best"` since `load_best_model_at_end` requires `eval_strategy == save_strategy` but `eval_strategy` doesn't have a `"best"` option.

This means that if we want to use `save_strategy = "best"` then we have to have `load_best_model_at_end = False`, which in turn means that when `save_strategy != "best"` and `load_best_model_at_end = True` then the `__post_init__` method of TrainingArguments is setting `metric_for_best_model` to `"loss"`. https://github.com/huggingface/transformers/blob/main/src/transformers/training_args.py#L1676:L1679

The modified docstring aims to add a bit more detail as to when the `metric_for_best_model` is set to a default of `"loss"`.

Reply:
should we also add `best` for `eval_strategy` then ?

Reply:
I feel like that might be a bit awkward since it sounds a bit awkward to perform evaluation at every best checkpoint.

Maybe we could check if `save_strategy == "best"` and then bypasses the `eval_strategy == save_strategy` condition? That would mean that [here](https://github.com/huggingface/transformers/blob/5c75087aeee7081025370e10d1f571a11600f1ae/src/transformers/training_args.py#L1639C9-L1644C18) we would change the code to:

```Python
if self.load_best_model_at_end and self.save_strategy != "best":
    if self.eval_strategy != self.save_strategy:
        raise ValueError(
            "--load_best_model_at_end requires the save and eval strategy to match, but found\n- Evaluation "
            f"strategy: {self.eval_strategy}\n- Save strategy: {self.save_strategy}"
        )
```

I'm not 100% sure about the history of why `eval_strategy == save_strategy` but I'm assuming that it's a safe guard to prevent situations where we want to load the best model at the end of training but we never saved it because the two didn't match. If `save_strategy == "best"` I don't think we'd have that problem since saving is guaranteed to following evaluation.

I think this also means that we may have to either remove the default loss error we're raising or change it to a warning (https://github.com/huggingface/transformers/pull/31817#discussion_r1806713897).

Reply:
Let's do that instead and add a comment to explain why it is fine to have this: 
```python
if self.load_best_model_at_end:
    if self.eval_strategy != self.save_strategy or self.save_strategy != "best":
```

Reply:
Sounds good to me for the new condition, just add a comment to explain why we don't need to perform the check when self.save_strategy == "best". Also, which default loss error you are talking about ? I'm not sure why we need to remove it.